src/backend/distributed/commands/multi_copy.c

```diff
@@ -1190,8 +1383,69 @@ void
 AppendCopyBinaryFooters(CopyOutState footerOutputState)
 {
 	int16 negative = -1;
+	MemoryContext oldContext = MemoryContextSwitchTo(footerOutputState->rowcontext);
 
 	CopySendInt16(footerOutputState, negative);
+
+	MemoryContextSwitchTo(oldContext);
+}
+
+
+/*
+ * StartCopyToNewShard creates a new shard and related shard placements, opens
+ * connections to shard placements, and starts copy command by sending copy
+ * headers.
+ */
+static void
+StartCopyToNewShard(CopyOutState copyOutState, ShardConnections *shardConnections,
+					Oid relationId, CopyStmt *copyStatement)
+{
+	char *relationName = get_rel_name(relationId);
+	text *relationNameText = cstring_to_text(relationName);
+	Datum relationNameDatum = PointerGetDatum(relationNameText);
+	Datum shardIdDatum = DirectFunctionCall1(master_create_empty_shard,
+											 relationNameDatum);
+
+	int64 shardId = DatumGetInt64(shardIdDatum);
+	shardConnections->shardId = shardId;
+
+	list_free_deep(shardConnections->connectionList);
+	shardConnections->connectionList = NIL;
+
+	/* connect to shards placements and start transactions */
+	OpenCopyTransactions(copyStatement, shardConnections);
+
+	/* send binary headers to shard placements */
+	resetStringInfo(copyOutState->fe_msgbuf);
```

Comment:
I wrapped header and footer parts as separate functions.

I also think about putting row copy part into a function, but it becomes a 3 lines function with 6 parameters. Therefore, I leave it as it is. Function would be like following:

```sql
static void
SendCopyRow(CopyOutState copyOutState, List *connectionList, Datum *columnValues,
			bool *columnNulls, TupleDesc tupleDescriptor, FmgrInfo *columnOutputFunctions)
{
	resetStringInfo(copyOutState->fe_msgbuf);
	AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
					  copyOutState, columnOutputFunctions);
	SendCopyDataToAll(copyOutState->fe_msgbuf, connectionList);
}
```